File: lib/llm/router.ts
/**
 * LLM router — picks an AI SDK provider based on configured keys.
 *
 * Order of preference: Anthropic Claude → OpenAI GPT → no-key stub.
 * The router is the single import point — agents never talk to providers directly.
 */
import { anthropic } from "@ai-sdk/anthropic";
import { openai } from "@ai-sdk/openai";
import type { LanguageModelV1 } from "ai";

export type RoutingPolicy = "quality" | "balanced" | "cheap" | "latency";

const DEFAULTS: Record<RoutingPolicy, { anthropic?: string; openai?: string }> = {
  quality: { anthropic: "claude-opus-5", openai: "gpt-4o" },
  balanced: { anthropic: "claude-sonnet-5", openai: "gpt-4o-mini" },
  cheap: { anthropic: "claude-haiku-4-5", openai: "gpt-4o-mini" },
  latency: { anthropic: "claude-haiku-4-5", openai: "gpt-4o-mini" },
};

export function pickModel(policy: RoutingPolicy = "balanced"): LanguageModelV1 {
  const choice = DEFAULTS[policy];

  if (process.env.ANTHROPIC_API_KEY && choice.anthropic) {
    return anthropic(choice.anthropic);
  }
  if (process.env.OPENAI_API_KEY && choice.openai) {
    return openai(choice.openai);
  }
  throw new Error(
    "No LLM provider configured. Set ANTHROPIC_API_KEY or OPENAI_API_KEY in your env.",
  );
}

export function hasAnyProvider(): boolean {
  return Boolean(process.env.ANTHROPIC_API_KEY || process.env.OPENAI_API_KEY);
}

// Approximate pricing in EUR per 1M tokens.
const PRICING: Record<string, { in: number; out: number }> = {
  "claude-opus-5": { in: 4.6, out: 23 },
  "claude-sonnet-5": { in: 2.8, out: 13.8 },
  "claude-haiku-4-5": { in: 0.92, out: 4.6 },
  "gpt-4o": { in: 2.3, out: 9 },
  "gpt-4o-mini": { in: 0.13, out: 0.55 },
};

export function estimateCostEur(model: string, tokensIn: number, tokensOut: number): number {
  const p = PRICING[model] ?? { in: 0, out: 0 };
  return (tokensIn * p.in + tokensOut * p.out) / 1_000_000;
}

File: legacy/backend/src/axion/llm/router.py
"""LLM routing layer.

Routes a request to the best available model based on:
- task type (sales / legal / engineering / general)
- latency vs. quality preference
- provider availability
- explicit override

Falls back through the chain on provider error.
"""
from __future__ import annotations

import enum
from dataclasses import dataclass
from typing import Any

import structlog

from axion.config import get_settings
from axion.llm.providers.anthropic_provider import AnthropicProvider
from axion.llm.providers.base import LLMProvider, NullProvider
from axion.llm.providers.mistral_provider import MistralProvider
from axion.llm.providers.openai_provider import OpenAIProvider

log = structlog.get_logger(__name__)
settings = get_settings()


class RoutingPolicy(str, enum.Enum):
    QUALITY = "quality"      # best model regardless of cost
    BALANCED = "balanced"    # default
    CHEAP = "cheap"          # minimize cost
    LATENCY = "latency"      # minimize round-trip


@dataclass
class ModelChoice:
    provider: str
    model: str


_ROUTING: dict[RoutingPolicy, list[ModelChoice]] = {
    RoutingPolicy.QUALITY: [
        ModelChoice("anthropic", "claude-opus-4-7"),
        ModelChoice("openai", "gpt-4o"),
        ModelChoice("mistral", "mistral-large-2"),
    ],
    RoutingPolicy.BALANCED: [
        ModelChoice("anthropic", "claude-sonnet-4-7"),
        ModelChoice("openai", "gpt-4o-mini"),
        ModelChoice("mistral", "mistral-medium"),
    ],
    RoutingPolicy.CHEAP: [
        ModelChoice("openai", "gpt-4o-mini"),
        ModelChoice("anthropic", "claude-haiku-4-7"),
        ModelChoice("mistral", "mistral-small"),
    ],
    RoutingPolicy.LATENCY: [
        ModelChoice("anthropic", "claude-haiku-4-7"),
        ModelChoice("openai", "gpt-4o-mini"),
    ],
}


class LLMRouter:
    """Routes completion requests to the appropriate provider."""

    def __init__(self) -> None:
        self._providers: dict[str, LLMProvider] = {
            "anthropic": AnthropicProvider() if settings.anthropic_api_key.get_secret_value() else NullProvider("anthropic"),
            "openai": OpenAIProvider() if settings.openai_api_key.get_secret_value() else NullProvider("openai"),
            "mistral": MistralProvider() if settings.mistral_api_key.get_secret_value() else NullProvider("mistral"),
        }

    async def complete(
        self,
        *,
        system: str,
        user: str,
        history: list[dict] | None = None,
        tools: list[dict] | None = None,
        max_tokens: int = 2048,
        policy: RoutingPolicy = RoutingPolicy.BALANCED,
        force_model: str | None = None,
    ) -> dict[str, Any]:
        """Run a completion, falling back through the routing chain."""
        if force_model:
            for provider_name, provider in self._providers.items():
                if force_model in provider.supported_models():
                    return await self._call(
                        provider_name, force_model, system, user, history, tools, max_tokens
                    )

        chain = _ROUTING.get(policy, _ROUTING[RoutingPolicy.BALANCED])
        last_error: Exception | None = None
        for choice in chain:
            try:
                return await self._call(
                    choice.provider, choice.model, system, user, history, tools, max_tokens
                )
            except Exception as e:  # noqa: BLE001
                last_error = e
                log.warning(
                    "llm_provider_failed",
                    provider=choice.provider,
                    model=choice.model,
                    error=str(e),
                )
                continue

        # All providers failed: return a no-op shape so agents degrade gracefully
        return {
            "text": f"[LLM unavailable: {last_error}]",
            "tokens_in": 0,
            "tokens_out": 0,
            "cost_eur": 0.0,
            "model": "none",
            "tool_calls": [],
        }

    async def _call(
        self,
        provider_name: str,
        model: str,
        system: str,
        user: str,
        history: list[dict] | None,
        tools: list[dict] | None,
        max_tokens: int,
    ) -> dict[str, Any]:
        provider = self._providers[provider_name]
        return await provider.complete(
            system=system,
            user=user,
            history=history or [],
            tools=tools or [],
            max_tokens=max_tokens,
            model=model,
        )
